### DISCUSSION TITLE
Is it possible to run UERANSIM on arm32v7？

### DISCUSSION THREAD

**Discussion by leegdut** [aligungr/UERANSIM#448] [Q&A]
Hi.
UERANSIM is awesome!!!
I have successfully run open5gs and UERANSIM on ubuntu.

And I want to run UERANSIM on my device(arm32v7, 32bit).
So I want to know it's possible to make UERANSIM for arm32v7?

I have modify the CMakeLists.txt to specify the compiler, but there are many errors.
Is there any guidance or reference?
Please let me know. Thank you so much.

Best regards

**Comment by aligungr** [comment]
Hi, UERANSIM supports only **64-bit** architectures for now. You can use ARM, AMD or any other 64-bit arc.

**Comment by leegdut** [comment]
Thanks for your reply. And I will try to migrate UERANSIM to armv7-32bit.

**Reply by aligungr** [reply]
@leegdut Please don't waste time for migrating UERANSIM to 32-bit. Because it has very low level codes and really complex procedure. I'll migrate it to 32-bit soon.

**Reply by leegdut** [reply]
Thank you so much. May I ask when the migration will be completed? Thanks！

**Reply by aligungr** [reply]
@leegdut I'm planning to complete today.

**Reply by leegdut** [reply]
It's amazing. Thank you so much.

**Reply by aligungr** [reply]
Now it's done.

But I could not test it, because I have no access to a 32-bit machine.

Please let me know if it compiles and runs successfully.

**Reply by leegdut** [reply]
I will try now. It's use the same way to compile（https://github.com/aligungr/UERANSIM/wiki/Installation）? My toolchain is arm-linux-gnueabi-gcc and arm-linux-gnueabi-g++, should I specify the gcc and g++?

**Reply by aligungr** [reply]
Yes, it's the same way. Cmake should handle the rest.

**Reply by leegdut** [reply]
Direct make is OK. 
But I am not build the project on my 32-bit machine.  I need to cross compile the UERANSIM on my x86_64 ubuntu, then put the output binaries to my 32-bit machine.
So I modify the CMakeLists.txt to specify the compiler:

SET(CMAKE_C_COMPILER "arm-linux-gnueabi-gcc")
SET(CMAKE_CXX_COMPILER "arm-linux-gnueabi-g++")


Compile the program but have the following error：

[ 91%] Building CXX object src/lib/CMakeFiles/common-lib.dir/asn/ngap.cpp.o
In file included from /usr/lwj/resource/real_cross/UERANSIM/src/lib/asn/ngap.cpp:10:
/usr/lwj/resource/real_cross/UERANSIM/src/lib/asn/utils.hpp: In function 'void asn::SetBitStringInt(int, BIT_STRING_t&)':
/usr/lwj/resource/real_cross/UERANSIM/src/lib/asn/utils.hpp:91:44: error: call of overloaded 'NearDiv(unsigned int, int)' is ambiguous
   91 |     target.size = bits::NearDiv(BitCount, 8) / 8;
      |                                            ^
In file included from /usr/lwj/resource/real_cross/UERANSIM/src/utils/octet_view.hpp:11,
                 from /usr/lwj/resource/real_cross/UERANSIM/src/lib/asn/utils.hpp:28,
                 from /usr/lwj/resource/real_cross/UERANSIM/src/lib/asn/ngap.cpp:10:
/usr/lwj/resource/real_cross/UERANSIM/src/utils/bits.hpp:115:17: note: candidate: 'unsigned int bits::NearDiv(unsigned int, unsigned int)'
  115 | inline unsigned NearDiv(unsigned val, unsigned div)
      |                 ^~~~~~~
/usr/lwj/resource/real_cross/UERANSIM/src/utils/bits.hpp:120:12: note: candidate: 'int bits::NearDiv(int, int)'
  120 | inline int NearDiv(int val, int div)
      |            ^~~~~~~
In file included from /usr/lwj/resource/real_cross/UERANSIM/src/lib/asn/ngap.cpp:10:
/usr/lwj/resource/real_cross/UERANSIM/src/lib/asn/utils.hpp: In function 'void asn::SetBitStringLong(int64_t, BIT_STRING_t&)':
/usr/lwj/resource/real_cross/UERANSIM/src/lib/asn/utils.hpp:104:44: error: call of overloaded 'NearDiv(unsigned int, int)' is ambiguous
  104 |     target.size = bits::NearDiv(BitCount, 8) / 8;
      |                                            ^
In file included from /usr/lwj/resource/real_cross/UERANSIM/src/utils/octet_view.hpp:11,
                 from /usr/lwj/resource/real_cross/UERANSIM/src/lib/asn/utils.hpp:28,
                 from /usr/lwj/resource/real_cross/UERANSIM/src/lib/asn/ngap.cpp:10:
/usr/lwj/resource/real_cross/UERANSIM/src/utils/bits.hpp:115:17: note: candidate: 'unsigned int bits::NearDiv(unsigned int, unsigned int)'
  115 | inline unsigned NearDiv(unsigned val, unsigned div)
      |                 ^~~~~~~
/usr/lwj/resource/real_cross/UERANSIM/src/utils/bits.hpp:120:12: note: candidate: 'int bits::NearDiv(int, int)'
  120 | inline int NearDiv(int val, int div)
      |            ^~~~~~~
make[3]: *** [src/lib/CMakeFiles/common-lib.dir/build.make:160: src/lib/CMakeFiles/common-lib.dir/asn/ngap.cpp.o] Error 1
make[3]: Leaving directory '/usr/lwj/resource/real_cross/UERANSIM/cmake-build-release'
make[2]: *** [CMakeFiles/Makefile2:487: src/lib/CMakeFiles/common-lib.dir/all] Error 2
make[2]: Leaving directory '/usr/lwj/resource/real_cross/UERANSIM/cmake-build-release'
make[1]: *** [Makefile:91: all] Error 2
make[1]: Leaving directory '/usr/lwj/resource/real_cross/UERANSIM/cmake-build-release'
make: *** [makefile:12: build] Error 2

**Reply by aligungr** [reply]
Nice catch. Actually that was a different error, and I fixed just now. Please try again with the latest code.

**Reply by leegdut** [reply]
Thanks! The above error has been solved.

And the new error as follow:
/usr/lwj/resource/new_cross/UERANSIM/src/lib/asn/utils.cpp: In function 'uint64_t asn::GetUnsigned64(const INTEGER_t&)':
/usr/lwj/resource/new_cross/UERANSIM/src/lib/asn/utils.cpp:132:36: error: cannot convert 'uint64_t*' {aka 'long long unsigned int*'} to 'long unsigned int*'
  132 |     if (asn_INTEGER2ulong(&source, &res) != 0)
      |                                    ^~~~
      |                                    |
      |                                    uint64_t* {aka long long unsigned int*}
In file included from /usr/lwj/resource/new_cross/UERANSIM/src/asn/asn1c/NativeInteger.h:16,
                 from /usr/lwj/resource/new_cross/UERANSIM/src/asn/asn1c/NativeEnumerated.h:15,
                 from /usr/lwj/resource/new_cross/UERANSIM/src/lib/asn/utils.hpp:20,
                 from /usr/lwj/resource/new_cross/UERANSIM/src/lib/asn/utils.cpp:9:
/usr/lwj/resource/new_cross/UERANSIM/src/asn/asn1c/INTEGER.h:75:58: note:   initializing argument 2 of 'int asn_INTEGER2ulong(const INTEGER_t*, long unsigned int*)'
   75 | int asn_INTEGER2ulong(const INTEGER_t *i, unsigned long *l);
      |                                           ~~~~~~~~~~~~~~~^
/usr/lwj/resource/new_cross/UERANSIM/src/lib/asn/utils.cpp: In function 'int64_t asn::GetSigned64(const INTEGER_t&)':
/usr/lwj/resource/new_cross/UERANSIM/src/lib/asn/utils.cpp:143:35: error: cannot convert 'int64_t*' {aka 'long long int*'} to 'long int*'
  143 |     if (asn_INTEGER2long(&source, &res) != 0)
      |                                   ^~~~
      |                                   |
      |                                   int64_t* {aka long long int*}
In file included from /usr/lwj/resource/new_cross/UERANSIM/src/asn/asn1c/NativeInteger.h:16,
                 from /usr/lwj/resource/new_cross/UERANSIM/src/asn/asn1c/NativeEnumerated.h:15,
                 from /usr/lwj/resource/new_cross/UERANSIM/src/lib/asn/utils.hpp:20,
                 from /usr/lwj/resource/new_cross/UERANSIM/src/lib/asn/utils.cpp:9:
/usr/lwj/resource/new_cross/UERANSIM/src/asn/asn1c/INTEGER.h:74:48: note:   initializing argument 2 of 'int asn_INTEGER2long(const INTEGER_t*, long int*)'
   74 | int asn_INTEGER2long(const INTEGER_t *i, long *l);
      |                                          ~~~~~~^
make[3]: *** [src/lib/CMakeFiles/common-lib.dir/build.make:188: src/lib/CMakeFiles/common-lib.dir/asn/utils.cpp.o] Error 1
make[3]: Leaving directory '/usr/lwj/resource/new_cross/UERANSIM/cmake-build-release'
make[2]: *** [CMakeFiles/Makefile2:487: src/lib/CMakeFiles/common-lib.dir/all] Error 2
make[2]: Leaving directory '/usr/lwj/resource/new_cross/UERANSIM/cmake-build-release'
make[1]: *** [Makefile:91: all] Error 2
make[1]: Leaving directory '/usr/lwj/resource/new_cross/UERANSIM/cmake-build-release'
make: *** [makefile:12: build] Error 2


I install the cross compiler on my ubuntu as follow:
apt-get install gcc-arm-linux-gnueabi
apt-get install g++-arm-linux-gnueabihf

And modify the CMakeLists.txt to specify the compiler:
SET(CMAKE_C_COMPILER "arm-linux-gnueabi-gcc")
SET(CMAKE_CXX_COMPILER "arm-linux-gnueabi-g++")

Maybe you can specify the compiler to make the project.I think it will help to fix the errors more effective.

Best Regards.

**Reply by aligungr** [reply]
I fixed the new error, thanks. Waiting for the next one :)

**Reply by leegdut** [reply]
👍 ! The above error has been solved.
And the first error is :

make[3]: Entering directory '/usr/lwj/resource/new_cross/UERANSIM/cmake-build-release'
[ 93%] Building CXX object CMakeFiles/nr-cli.dir/src/cli.cpp.o
[ 93%] Linking CXX executable nr-cli
/usr/lib/gcc-cross/arm-linux-gnueabi/9/../../../../arm-linux-gnueabi/bin/ld: cannot find -lsctp
collect2: error: ld returned 1 exit status
make[3]: *** [CMakeFiles/nr-cli.dir/build.make:103: nr-cli] Error 1
make[3]: Leaving directory '/usr/lwj/resource/new_cross/UERANSIM/cmake-build-release'
make[2]: *** [CMakeFiles/Makefile2:237: CMakeFiles/nr-cli.dir/all] Error 2
make[2]: Leaving directory '/usr/lwj/resource/new_cross/UERANSIM/cmake-build-release'
make[1]: *** [Makefile:91: all] Error 2
make[1]: Leaving directory '/usr/lwj/resource/new_cross/UERANSIM/cmake-build-release'
make: *** [makefile:12: build] Error 2

I think it can't find the 32-bit sctp library, so I fix the proble as follow, but i don't know if it is correct or not.

tar -xzxf lksctp-tools-1.0.11.tar.gz
./configure  --prefix=$prefix --host=arm  CC=arm-linux-gnueabi-gcc
make

then modify the CMakeLists.txt:
LINK_DIRECTORIES(/usr/lwj/resource/sctp/lksctp-tools-1.0.11/src/lib/.libs)

And the next error is：
make[3]: Entering directory '/usr/lwj/resource/new_cross/UERANSIM/cmake-build-release'
[ 93%] Linking CXX executable nr-cli
/usr/lib/gcc-cross/arm-linux-gnueabi/9/../../../../arm-linux-gnueabi/bin/ld: src/utils/libutils.a(random.cpp.o): in function `Random::Random()':
random.cpp:(.text+0x40): undefined reference to `__atomic_fetch_add_8'
collect2: error: ld returned 1 exit status
make[3]: *** [CMakeFiles/nr-cli.dir/build.make:103: nr-cli] Error 1
make[3]: Leaving directory '/usr/lwj/resource/new_cross/UERANSIM/cmake-build-release'
make[2]: *** [CMakeFiles/Makefile2:237: CMakeFiles/nr-cli.dir/all] Error 2
make[2]: Leaving directory '/usr/lwj/resource/new_cross/UERANSIM/cmake-build-release'
make[1]: *** [Makefile:91: all] Error 2
make[1]: Leaving directory '/usr/lwj/resource/new_cross/UERANSIM/cmake-build-release'
make: *** [makefile:12: build] Error 2

**Reply by aligungr** [reply]
See the link:
 https://github.com/redis/redis/issues/6275

`-latomic` flag may fix the problem.

**Reply by leegdut** [reply]
Unfortunately, the problem was not resolved after I added -latomic flag. I added -latomic flag to CMakeLists.txt like this:
![image](https://user-images.githubusercontent.com/42730971/144775947-235130c1-6ce2-4566-94ab-83c5c290045d.png)

Maybe you can repeat the error by installing the cross-compilation chain. Thanks！

**Reply by DiogoCruz40** [reply]
Hey @leegdut, 
Did you manage to resolve this problem? I used the same flag as you and am experiencing the same problem.

**Reply by DiogoCruz40** [reply]
By installing Ubuntu Server 22.04 LTS (32 bits) on the Rapsberry PI and following the installation instructions in the UERANSIM official repo, I was able to resolve the issue.

**Comment by leegdut** [comment]
Hi,I just run UERANSIM on my arm32 device and it's very close to being able to connect with open5gs.
But there is still an error as follows：

/storage/ue # ./nr-gnb -c new_open5gs-gnb.yaml &                                
[1]-  Done                       ./nr-gnb -c new_open5gs-gnb.yaml               
/storage/ue # UERANSIM v3.2.4                                                   
[2000-08-21 20:14:31.022] [sctp] [info] Trying to establish SCTP connection... (
70.32.36.53:38412)                                                              
[2000-08-21 20:14:31.043] [sctp] [info] SCTP connection established (70.32.36.53
:38412)                                                                         
[2000-08-21 20:14:31.043] [sctp] [debug] SCTP association setup ascId[4]        
[2000-08-21 20:14:31.044] [ngap] [debug] Sending NG Setup Request               
[2000-08-21 20:14:31.046] [ngap] [debug] NG Setup Response received             
[2000-08-21 20:14:31.046] [ngap] [info] NG Setup procedure is successful        


/storage/ue # ./nr-ue -c new_open5gs-ue.yaml &                                  
/storage/ue # UERANSIM v3.2.4                                                   
[2000-08-21 20:14:53.718] [nas] [info] UE switches to state [MM-DEREGISTERED/PLM
N-SEARCH]                                                                       
[2000-08-21 20:14:53.720] [rrc] [debug] UE[1] new signal detected               
[2000-08-21 20:14:53.720] [rrc] [debug] New signal detected for cell[1], total [
1] cells in coverage                                                            
[2000-08-21 20:14:53.721] [nas] [info] Selected plmn[001/01]                    
[2000-08-21 20:14:53.721] [rrc] [info] Selected cell plmn[001/01] tac[1] categor
y[SUITABLE]                                                                     
[2000-08-21 20:14:53.722] [nas] [info] UE switches to state [MM-DEREGISTERED/PS]
[2000-08-21 20:14:53.722] [nas] [info] UE switches to state [MM-DEREGISTERED/NOR
MAL-SERVICE]                                                                    
[2000-08-21 20:14:53.722] [nas] [debug] Initial registration required due to [MM
-DEREG-NORMAL-SERVICE]                                                          
[2000-08-21 20:14:53.722] [nas] [debug] UAC access attempt is allowed for identi
ty[0], category[MO_sig]                                                         
[2000-08-21 20:14:53.722] [nas] [debug] Sending Initial Registration            
[2000-08-21 20:14:53.722] [nas] [info] UE switches to state [MM-REGISTER-INITIAT
ED]                                                                             
[2000-08-21 20:14:53.722] [rrc] [debug] Sending RRC Setup Request               
[2000-08-21 20:14:53.723] [rrc] [info] RRC Setup for UE[1]                      
[2000-08-21 20:14:53.724] [rrc] [info] RRC connection established               
[2000-08-21 20:14:53.724] [rrc] [info] UE switches to state [RRC-CONNECTED]     
[2000-08-21 20:14:53.724] [ngap] [debug] Initial NAS message received from UE[1]
[2000-08-21 20:14:53.725] [nas] [info] UE switches to state [CM-CONNECTED]      
[2000-08-21 20:14:53.744] [nas] [debug] Authentication Request received         
[2000-08-21 20:14:53.745] [ngap] [error] NGAP APER encoding failed              
                                                                                
/storage/ue #                                                                   
/storage/ue # [2000-08-21 20:14:59.748] [nas] [debug] Authentication Request rec
eived                                                                           
[2000-08-21 20:14:59.748] [nas] [error] Sending Authentication Failure with caus
e [NGKSI_ALREADY_IN_USE]                                                        
[2000-08-21 20:14:59.749] [ngap] [error] NGAP APER encoding failed              
                                                                                
/storage/ue # [2000-08-21 20:15:05.754] [nas] [debug] Authentication Request rec
eived                                                                           
[2000-08-21 20:15:05.754] [nas] [error] Sending Authentication Failure with caus
e [NGKSI_ALREADY_IN_USE]                                                        
[2000-08-21 20:15:05.755] [ngap] [error] NGAP APER encoding failed              
[2000-08-21 20:15:08.857] [nas] [debug] NAS timer[3510] expired [1]             
[2000-08-21 20:15:08.857] [nas] [info] UE switches to state [MM-DEREGISTERED/PS]
[2000-08-21 20:15:08.857] [nas] [info] UE switches to state [5U2-NOT-UPDATED]   
[2000-08-21 20:15:08.857] [nas] [info] Performing local release of NAS connectio
n                                                                               
[2000-08-21 20:15:08.858] [nas] [info] UE switches to state [MM-DEREGISTERED/ATT
EMPTING-REGISTRATION]                                                           
[2000-08-21 20:15:08.858] [rrc] [info] UE switches to state [RRC-IDLE]          
[2000-08-21 20:15:08.858] [nas] [info] UE switches to state [CM-IDLE]           
[2000-08-21 20:15:09.743] [rrc] [debug] UE[2] new signal detected               
[2000-08-21 20:15:11.202] [rls] [debug] UE[1] signal lost                       



and the open5gs shows:
root@VOLT2:/usr/lwj/open5gs/open5gs# 01/20 12:14:07.176: [amf] INFO: InitialUEMessage (../src/amf/ngap-handler.c:361)
01/20 12:14:07.178: [amf] INFO: [Added] Number of gNB-UEs is now 1 (../src/amf/context.c:2036)
01/20 12:14:07.180: [amf] INFO:     RAN_UE_NGAP_ID[3] AMF_UE_NGAP_ID[3] TAC[1] CellID[0x10] (../src/amf/ngap-handler.c:497)
01/20 12:14:07.182: [amf] INFO: [suci-0-001-01-0000-0-0-0000000001] Unknown UE by SUCI (../src/amf/context.c:1388)
01/20 12:14:07.183: [amf] INFO: [Added] Number of AMF-UEs is now 1 (../src/amf/context.c:1187)
01/20 12:14:07.183: [gmm] INFO: Registration request (../src/amf/gmm-sm.c:134)
01/20 12:14:07.183: [gmm] INFO: [suci-0-001-01-0000-0-0-0000000001]    SUCI (../src/amf/gmm-handler.c:156)
01/20 12:14:32.407: [amf] INFO: InitialUEMessage (../src/amf/ngap-handler.c:361)
01/20 12:14:32.407: [amf] INFO: [Added] Number of gNB-UEs is now 2 (../src/amf/context.c:2036)
01/20 12:14:32.407: [amf] INFO:     RAN_UE_NGAP_ID[4] AMF_UE_NGAP_ID[4] TAC[1] CellID[0x10] (../src/amf/ngap-handler.c:497)
01/20 12:14:32.407: [amf] INFO: [suci-0-001-01-0000-0-0-0000000001] known UE by SUCI (../src/amf/context.c:1386)
01/20 12:14:32.407: [gmm] WARNING: Registration request (../src/amf/gmm-sm.c:573)
01/20 12:14:32.407: [gmm] INFO: [suci-0-001-01-0000-0-0-0000000001]    SUCI (../src/amf/gmm-handler.c:156)
01/20 12:14:32.407: [amf] WARNING: GUTI has already been allocated (../src/amf/context.c:1049)
01/20 12:15:02.415: [amf] WARNING: Implicit NG release (../src/amf/amf-sm.c:713)
01/20 12:15:02.415: [amf] WARNING:     RAN_UE_NGAP_ID[3] AMF_UE_NGAP_ID[3] (../src/amf/amf-sm.c:714)
01/20 12:15:02.415: [amf] INFO: UE Context Release [Action:1] (../src/amf/ngap-handler.c:1398)
01/20 12:15:02.415: [amf] INFO:     RAN_UE_NGAP_ID[3] AMF_UE_NGAP_ID[3] (../src/amf/ngap-handler.c:1399)
01/20 12:15:02.415: [amf] INFO: [Removed] Number of gNB-UEs is now 1 (../src/amf/context.c:2042)
01/20 12:15:02.434: [gmm] WARNING: [suci-0-001-01-0000-0-0-0000000001] Retransmission failed. Stop retransmission (../src/amf/gmm-sm.c:613)
01/20 12:15:02.434: [amf] WARNING: [suci-0-001-01-0000-0-0-0000000001] Authentication reject (../src/amf/nas-path.c:342)
01/20 12:15:32.441: [amf] WARNING: Implicit NG release (../src/amf/amf-sm.c:713)
01/20 12:15:32.443: [amf] WARNING:     RAN_UE_NGAP_ID[4] AMF_UE_NGAP_ID[4] (../src/amf/amf-sm.c:714)
01/20 12:15:32.445: [amf] INFO: UE Context Release [Action:3] (../src/amf/ngap-handler.c:1398)
01/20 12:15:32.447: [amf] INFO:     RAN_UE_NGAP_ID[4] AMF_UE_NGAP_ID[4] (../src/amf/ngap-handler.c:1399)
01/20 12:15:32.448: [amf] INFO:     SUCI[suci-0-001-01-0000-0-0-0000000001] (../src/amf/ngap-handler.c:1402)
01/20 12:15:32.448: [amf] INFO: [Removed] Number of gNB-UEs is now 0 (../src/amf/context.c:2042)
01/20 12:15:32.449: [amf] INFO: [Removed] Number of AMF-UEs is now 0 (../src/amf/context.c:1271)

Do you know what cause the errors?  Waiting for your reply.
Best Regards.

**Comment by aligungr** [comment]
Hi

Please try the latest version, it may fix the error. Indeed 3.2.4 is an old version.

BR
